File: AtCoder/abc/187/f.py
n, m = map(int, input().split())
dp = [n] * (1 << n) 
adj = [[False] * n for _ in range(n)]
for i in range(n):
    adj[i][i] = True
for _ in range(m):
    u, v = map(lambda x: int(x) - 1, input().split())
    adj[u][v] = True
    adj[v][u] = True
for mask in range(1, 1 << n):
    l = []
    for i in range(n):
        if (mask >> i) & 1:
            l.append(i)
    for a in l:
        for b in l:
            if not adj[a][b]:
                break
        else:
            continue
        break
    else:
        dp[mask] = 1
        continue
    sub = (mask - 1) & mask
    while sub:
        dp[mask] = min(dp[mask], dp[sub] + dp[mask ^ sub])
        sub = (sub - 1) & mask
print(dp[-1])

File: AtCoder/abc/187/f.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false), cin.tie(nullptr);
    int n, m;
    cin >> n >> m;
    vector<vector<bool>> a(n, vector<bool>(n));
    vector<int> dp(1 << n, 0);
    for (int i = 0, u, v; i < m; i++) {
        cin >> u >> v;
        --u, --v;
        a[u][v] = a[v][u] = true;
    }
    dp[0] = 0;
    auto is_clique = [&](int mask) -> bool {
        for (int i = 0; i < n; i++) {
            if (!(mask & (1 << i))) continue;
            for (int j = 0; j < n; j++) {
                if (i == j || !(mask & (1 << j))) continue;
                if (!a[i][j]) return false;
            }
        }
        return true;
    };
    for (int mask = 1; mask < (1 << n); mask++) {
        if (__builtin_popcount(mask) == 1 || is_clique(mask)) {
            dp[mask] = 1;
            continue;
        }
        dp[mask] = int(1e9);
        for (int sub = mask; sub; sub = (sub - 1) & mask) {
            dp[mask] = min(dp[mask], dp[sub] + dp[mask ^ sub]);
        }
    }
    cout << dp[(1 << n) - 1] << '\n';
    return 0;
}
